### DISCUSSION TITLE
Using actions in Storybook 8 and Svelte 4

### DISCUSSION THREAD

**Discussion by alexlibby** [storybookjs/storybook#27040] [Help]
### Summary

I'm writing a basic component library with several components, in Svelte 4.

While components render OK, I'm really struggling to define actions such as onInput or onClicked using the new actions available in Storybook 8. For example, I have an Input component that looks like this:

```
<svelte:options customElement="garnet-input" />

<script>
	export let label = "Label:";
	export let placeholder = "";
	export let disabled = false;
	export let inputName = "";
	export let fieldID = "";
	export let email = "";

	export let value = undefined;

  function handleChange(event) {
    email = event.target.value;
  }
</script>

<div class="garnet">
	{#if label}
		<label for={fieldID}>{label}</label>
	{/if}
	<input
		type="text"
		id={fieldID}
		name={inputName}
		{placeholder}
		{disabled}
		bind:value
		on:input={handleChange}
		{...$$restProps}
	/>
</div>

<style>
_...removed for brevity_
</style>
```

In my `Input.stories.js` file, I have this:

```
import Input from "./Input.svelte";

export default {
  title: "Garnet UI Library/Basic Components/Input",
  component: Input,
  argTypes: {
    placeholder: "example text",
    label: "Text:",
    id: "default-variant",
    onChange: { action: "changed" },
  },
};

export const Default = {
  args: {
    placeholder: "example text",
    label: "Text:",
    id: "default-variant",
    onInput: () => console.log("this is an email field"),    
  },
};
```

When running up Storybook, I don't see anything happen at all - it's as though the actions don't exist. I've looked at other examples online but many don't yet use Storybook 8. I've also looked in GitHub issues and saw one about not including the `on:` when defining actionsData, but I couldn't get it to work. I'm keen to keep it simple for now - is anyone able to help point me in the right direction please? (Note: I am creating this component as a Web Component, hence the tag at the top - this shouldn't have any impact...I hope!)

**Comment by JReinhold** [comment]
You've removed the HTML markup from your example so it's hard to tell exactly what's going on. The important mental model to have is that Storybook can't do anything that a parent component to your Input wouldn't be able to do. So you need to expose that event from your component somehow. If you're using [declarative event forwarding](https://svelte.dev/docs/component-directives#on-eventname) that should already have been taken care of.

I've created a reproduction here showing multiple ways to solve your problem: https://stackblitz.com/~/edit/github-qjt72r?file=src%2Fstories%2FAutomaticInput.stories.js

If arg types inference works for your component, it should already "just work" and you don't need to manually attach any action listener.
If for some reason it doesn't work, you can attach them manually by supplying the `on` property in `render`. I understand your confusion because I don't think this manual attachment is documented anywhere. I've demonstrated two ways of manually attaching it depending on if you're in pre v8.1 or not.

With [Svelte 5's new `onX` event handling](https://svelte-5-preview.vercel.app/docs/event-handlers#component-events), the complexity around the `render` function shouldn't be needed anymore and you should be able to do `args: { onInput: fn() }`.
